---
title: "scRNAseq_MonocyteIntegration_1"
author: "Mike Caponegro"
date: "12/23/2019"
output: html_document
---

#Create LIGER object from scrath, from individual datasets. 

#Libraries
```{r message=FALSE, warning=FALSE}
library(Seurat)
library(ggplot2)
library(plotly)
library(SeuratWrappers)
library(affy)
library(matrixStats)
library(car)
library(dplyr)
library(ggrepel)
```

```{r LIGER 0.4.2}
devtools::install_github("jefferis/RANN@master-L1")
devtools::install_github("MacoskoLab/liger@v0.4.2", force = T)
packageVersion("liger") #should be '0.4.2"
```

#Counts/TPM from each dataset were combined into a list object, for Seurat/LIGER analysis 
```{r}
merge.seurat<-readRDS("./MyeloidIntegration_ListForLIGER.rds")
```

```{r}
##To know the current storage capacity
memory.limit()
## To increase the storage capacity
memory.limit(size=56000)
options(future.globals.maxSize = 4000 * 1024^2)
```

#I want to remove genes only expressed in a few cells
#Use min.cells in create suerat object, or row counts and remove left tails trnscirtps
```{r}
counts<-rowCounts(as.matrix(merge.seurat@assays$RNA@counts))
counts<-data.frame(counts=counts, row.names = rownames(merge.seurat))
hist(counts$counts)

counts7<-data.frame(counts=counts[counts$counts>=40,,drop=FALSE]) 
hist(counts7$counts)
nrow(counts)-nrow(counts7)

#remove lowly expressed transcripts from seurat object
gene.keep<-rownames(counts7)
merge.seurat<-merge.seurat[rownames(merge.seurat) %in% c(gene.keep),]

```

#Run QC
```{r}
merge.seurat[["percent.ribo"]]<-PercentageFeatureSet(merge.seurat, pattern = "RP")
merge.seurat[["percent.KCNQ1OT1"]] <- PercentageFeatureSet(merge.seurat, pattern = "KCNQ1OT1")

ERCC.WT.index <- grep(pattern = "^ERCC", x=rownames(merge.seurat), value = FALSE) 
percent.ERCC.WT <- Matrix::colSums(merge.seurat[ERCC.WT.index, ])/Matrix::colSums(merge.seurat)
merge.seurat@meta.data$ERCC <- data.frame(percent.ERCC = percent.ERCC.WT)

VlnPlot(merge.seurat, features = c("percent.mt", "percent.ribo", "percent.KCNQ1OT1", "ERCC"), group.by = "orig.ident")

#Remove ERCC from count.data
merge.seurat <- merge.seurat[-ERCC.WT.index, ]

#New ERCC
ERCC.WT.index <- grep(pattern = "^ERCC", x=rownames(merge.seurat), value = FALSE) 
percent.ERCC.WT <- Matrix::colSums(merge.seurat[ERCC.WT.index, ])/Matrix::colSums(merge.seurat)
merge.seurat@meta.data$ERCC.rm <- data.frame(percent.ERCC = percent.ERCC.WT)

VlnPlot(merge.seurat, features = c("ERCC.rm"), group.by = "orig.ident")

VlnPlot(merge.seurat, features = c("nFeature_RNA", "nCount_RNA"), group.by = "orig.ident")

```

#Use LIGER with Seurat Wrapper
```{r}
merge.seurat.liger <- NormalizeData(merge.seurat)
merge.seurat.liger <- FindVariableFeatures(merge.seurat.liger, selection.method = "vst", nfeatures = 3000)
merge.seurat.liger <- ScaleData(merge.seurat.liger, split.by = "orig.ident", do.center = FALSE)

merge.seurat.liger <- RunOptimizeALS(merge.seurat.liger, k = 13, lambda = 8, split.by = "orig.ident")

merge.seurat.liger <- RunQuantileAlignSNF(merge.seurat.liger, split.by = "orig.ident")

merge.seurat.liger <- RunUMAP(merge.seurat.liger, 
                                      dims = 1:ncol(merge.seurat.liger[["iNMF"]]), reduction = "iNMF",
                                      n.neighbors = 40, min.dist = 0.6)

DimPlot(merge.seurat.liger, group.by = c("labeled.ident2")) 
```

#Cell Cylce Scoring - after Normalization 
```{r}
merge.seurat.liger <- CellCycleScoring(
  object = merge.seurat.liger,
  g2m.features = cc.genes$g2m.genes,
  s.features = cc.genes$s.genes
)

VlnPlot(merge.seurat.liger, features = c("S.Score","G2M.Score"), group.by = "labeled.ident2")
```

#Rename Idents
```{r}

Idents(merge.seurat.genefilt.liger)<-'labeled.ident2'
tmp.cells <- WhichCells(object = merge.seurat.genefilt.liger, idents = c('GAM'))
merge.seurat.genefilt.liger <- SetIdent(merge.seurat.genefilt.liger, cells = tmp.cells, value = "Tumor Core GAM")
merge.seurat.genefilt.liger$labeled.ident2<-Idents(merge.seurat.genefilt.liger)


Idents(merge.seurat.genefilt.liger)<-'labeled.ident2'
tmp.cells <- WhichCells(object = merge.seurat.genefilt.liger, idents = c('Homeostatic Microglia', 'Repopulating Microglia',
                                                                'Tumor Core GAM', 'PGAM',
                                                                'Sankowski.GBM', 'Sankowski.Healthy'))
names<-colnames(merge.seurat.genefilt.liger)
tmp.cells.not<-names[!names %in% tmp.cells]

merge.seurat.genefilt.liger <- SetIdent(merge.seurat.genefilt.liger, cells = tmp.cells.not, value = "GAM - Other")

merge.seurat.genefilt.liger$labeled.ident2<-Idents(merge.seurat.genefilt.liger)

DimPlot(merge.seurat.genefilt.liger, group.by = "labeled.ident2", pt.size = 1, order=T)

```

#get feature laodings by ident
```{r}
Idents(merge.seurat.liger)<-"labeled.ident2"
tmp.sub<-subset(merge.seurat.liger, idents = "PGAM")

embed<-as.data.frame(tmp.sub@reductions[["iNMF_raw"]]@cell.embeddings)
load<-(tmp.sub@reductions[["iNMF_raw"]]@feature.loadings)

names(table(tmp.sub@meta.data[["labeled.ident2"]]))

#max row
test<-data.frame(max = rowMax(load), gene = rownames(load))
test<-test[order(test$max, decreasing = T),,drop=FALSE]

ggplot(test, aes(x = reorder(gene, max), y = max))+
  geom_point()+
  geom_label_repel(data = test[1:10,], aes(label = gene),
                  box.padding   = 0.35, 
                  point.padding = 0.5,
                  segment.color = 'grey50')+
  ggtitle("PGAM")

```

```{r}
heatcols<-CustomPalette(low = "blue", high = "red", mid = 'yellow', k = 100)
FeaturePlot(merge.seurat.liger, "IBSP", order = TRUE, cols = heatcols)

FeaturePlot(merge.seurat.liger, "percent.mt", order = TRUE)

Idents(merge.seurat.liger)<-"labeled.ident2"
merge.seurat.liger$labeled.ident2<-factor(merge.seurat.liger$labeled.ident2, 
                                          levels = c("Homeostatic Microglia", "Repopulating Microglia",
                                                     "Darmanis_Distant", "Tumor core GAM",
                                                     "Sankowski_GAM", 
                                                     "PGAM", "Sankowski_Aged", "NA"))

VlnPlot(merge.seurat.liger, features = c("MKI67", "CDK1", "CD36", "NES"), ncol = 2) #Repopulating Mg Markers
VlnPlot(merge.seurat.liger, features = c("ST6GAL1", "CX3CR1", "IPCEF1", "SELPLG", "DHRS9", "CD83"))
VlnPlot(merge.seurat.liger, features = c("ACY3", "BIN1", "GPR34", "ADGRG1", "NAV3", "EGR2"))
VlnPlot(merge.seurat.liger, features = c("CCL2", "CCL3", "CCL4", "IL1B", "TNF", "P2RY12"))

VlnPlot(merge.seurat.liger, features = c("NRP1", "VEGFA", "HIF1A", "TGFB1", "SEMA4A", "PLXNA1"))
VlnPlot(merge.seurat.liger, features = c("KRT17"))

VlnPlot(merge.seurat.liger, features = colnames(merge.seurat.liger@reductions$iNMF_raw@feature.loadings))

```


#Cluster the cells
```{r}

merge.seurat.liger <- RunPCA(merge.seurat.liger, verbose=FALSE, dims = 1:15) #use PCA instead of liger loadings??

merge.seurat.liger <- FindNeighbors(merge.seurat.liger)
merge.seurat.liger <- FindClusters(merge.seurat.liger, resolution = 0.5)

merge.seurat.liger$PCAclusters <- Idents(merge.seurat.liger) 


DimPlot(merge.seurat.liger, label = TRUE, split.by = "labeled.ident2", group.by = "seurat_clusters", ncol = 4)

DimPlot(merge.seurat.liger)
tmp.cells<-WhichCells(merge.seurat.liger, idents = c("MGH42", "MGH43", "MGH44", "MGH45",
                                          'MGH56', "MGH64", "MGH61", 'MGH57', 'MGH107'
                                          ))
tmp.cells<-WhichCells(merge.seurat.liger, idents = c("MGH115"))
DimPlot(merge.seurat.liger, cells.highlight = tmp.cells)
```

```{r}

markers4<-FindMarkers(ligerex, ident.1 = c(4), only.pos = TRUE)
#It looks like these cells are marked by "MTRN" genes
#they may have escaped QC in this way. Go back to Patinets and add tp percent.mt
markers0<-FindMarkers(ligerex, ident.1 = c(0), only.pos = TRUE)
#CCL2 CCL3 and CCL4 still mark these cells, albiet, only at about a 1.2 FC

```

```{r}
Idents(merge.seurat.liger)<-'labeled.ident2'

PGAM.markers<-FindMarkers(merge.seurat.liger, ident.1 = c('PGAM'), only.pos = TRUE)
PGAM.markers<-FindMarkers(ligerex, ident.1 = c(7), only.pos = TRUE)

markers <- FindAllMarkers(merge.seurat.liger, only.pos = TRUE)
top10 <- markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)

DoHeatmap(merge.seurat.liger, features = top10$gene) + NoLegend()

```

#3D projection
```{r}
library(plotly)

merge.seurat.liger3D<- RunUMAP(merge.seurat.liger, 
                               dims = 1:ncol(tmp.sub[["iNMF"]]), reduction = "iNMF",
                               n.neighbors = 10, min.dist = 0.2, n.components = 3L)

# Idents(merge.seurat.liger3D)<-'labeled.ident2'
# merge.seurat.liger3D<-subset(merge.seurat.liger3D, idents = c('PGAM', 'Homeostatic Microglia', 'Repopulating Microglia'))

# Extract tSNE information from Seurat Object
umap_1 <- merge.seurat.liger3D[["umap"]]@cell.embeddings[,1]
umap_2 <- merge.seurat.liger3D[["umap"]]@cell.embeddings[,2]
umap_3 <- merge.seurat.liger3D[["umap"]]@cell.embeddings[,3]

# Visualize what headings are called so that you can extract them to form a dataframe
#Embeddings(object = merge.seurat.integrated, reduction = "umap")

# Prepare a dataframe for cell plotting
plotting.data <- FetchData(object = merge.seurat.liger3D, vars = c("UMAP_1", "UMAP_2", "UMAP_3", "seurat_clusters", "labeled.ident2"))

# Make a column of row name identities (these will be your cell/barcode names)
plotting.data$label<-(colnames(merge.seurat.liger3D))


#cols = c('#809fff', "orangered", '#a300cc', '#ffcc66', '#cc0000', '#ff1ab3') #labeled,ident2 #labeled,ident2
cols <- c("#F8766D", "#619CFF", "#00BA38")
cols<-"grey"

# Plot your data
plot_ly(data = plotting.data, 
        x = ~UMAP_1, y = ~UMAP_2, z = ~UMAP_3, 
        color = ~labeled.ident2,
        colors = c(cols[1], cols[2], "#F8766D", "#00BA38", "#619CFF", cols[6], cols[7]),
        type = "scatter3d", 
        mode = "markers", 
        marker = list(size = 2, width=2), # controls size of points
        text=~label, #This is that extra column we made earlier for which we will use for cell ID
        hoverinfo="text") #When you visualize your plotly object, hovering your mouse pointer over a point shows cell names

```


#score with PGAM genes

```{r}

PGAM6<-read.csv("C:/Users/mcap9/Desktop/R Projects/RNAseq/Heatmaps/PGAM6_12.16.19.txt", stringsAsFactors = FALSE, header=FALSE)
PGAM6<-PGAM6$V1
othermy<-read.csv("C:/Users/mcap9/Desktop/R Projects/RNAseq/Heatmaps/data/OtherMyeloid.top100.txt", stringsAsFactors = FALSE, sep = '\t', header = TRUE)
othermy<-othermy[,1]

gene_lists<-list()

gene_lists$PGAM <- PGAM6
gene_lists$GAM <- othermy

```

```{r}

expression_matrix<-GetAssayData(merge.seurat.liger, slot = "counts")
expression_matrix<-log10(expression_matrix+1)
cell_metadata<-merge.seurat.liger@meta.data
gene_metadata<-as.data.frame(rownames(x= merge.seurat.liger))
rownames(gene_metadata)<-gene_metadata$`rownames(x = merge.seurat.integrated)`
colnames(gene_metadata)[1] <- "gene_short_name"


```


## Calculate Signature Scores for all cells in your seurat object
```{r}
# Just replace DF with your seurat Object name

data<-expression_matrix
ex<-as.matrix(data)
sampInfo<-data.frame(cell_metadata)
genes<-rownames(data)
featInfo<-data.frame(SYMBOLS=genes)
dataset <- list(ex = ex,
                   featInfo = featInfo,
                   sampInfo = sampInfo
                   )
rm(ex)
#===========================================================
expression <- as.matrix(dataset$ex)
dataset$featInfo$gene_short_name<-(dataset$featInfo$SYMBOLS)
genes <- dataset$featInfo
samples <- (dataset$sampInfo)
remove("dataset")
samples$note <- " "
for(i in names(gene_lists)){
  this_gene_list <- gene_lists[[i]]
  if(class(this_gene_list) %in% "data.frame"){
    tmp.symbols <- this_gene_list[,1]
    tmp.type <- as.character(this_gene_list[,2])
  }else{
    tmp.symbols <- this_gene_list
    tmp.type = rep(x = i,times = length(this_gene_list))
  }
  tmp.indexes <- match(table = genes$gene_short_name , x = tmp.symbols)
  tmp.symbols <- subset(tmp.symbols,!is.na(tmp.indexes))
  tmp.type <- subset(tmp.type,!is.na(tmp.indexes))
  tmp.indexes <- subset(tmp.indexes,!is.na(tmp.indexes))
  samples[i] <- colMeans((expression[tmp.indexes,,drop=FALSE]),na.rm = TRUE)
  genes[i] <- ""
  genes[tmp.indexes,i] <- tmp.type
  genes[i] <- as.factor(genes[[i]])
}
#===========================================================
rm("this_gene_list","i","tmp.symbols","tmp.type","tmp.indexes")


```

#Add scores to metadata
```{r}
merge.seurat.liger$PGAM_Score <- samples$PGAM
merge.seurat.liger$Tumor_Core_GAM_Score <- samples$GAM

heatcols<-CustomPalette(low = "blue2", high = "red2", mid = 'gold', k = 100)

FeaturePlot(object = merge.seurat.liger, reduction = "umap", features = c('Tumor_Core_GAM_Score',
                                                                          'PGAM_Score'),
            cols = heatcols, order = TRUE, pt.size = 0.5)  

VlnPlot(merge.seurat.liger, features = c('PGAM_Score'))
```


#pass to Monocole3
```{r}
library(monocle3)


gene_annotation <- as.data.frame(rownames(merge.seurat.liger@reductions[["iNMF_raw"]]@feature.loadings), row.names = rownames(merge.seurat.liger@reductions[["iNMF_raw"]]@feature.loadings))
colnames(gene_annotation) <- "gene_short_name"

cell_metadata <- as.data.frame(merge.seurat.liger@assays[["RNA"]]@counts@Dimnames[[2]], row.names = merge.seurat.liger@assays[["RNA"]]@counts@Dimnames[[2]])
colnames(cell_metadata) <- "barcode"

New_matrix <- merge.seurat.liger@assays[["RNA"]]@counts
New_matrix <- New_matrix[rownames(gene_annotation), ]
expression_matrix <- New_matrix

################ use all RNA counts?
# gene_annotation <- as.data.frame(rownames(merge.seurat.liger),
#                                  row.names = rownames(merge.seurat.liger))
# colnames(gene_annotation) <- "gene_short_name"
# 
# cell_metadata <- as.data.frame(merge.seurat.liger@assays[["RNA"]]@counts@Dimnames[[2]], row.names = merge.seurat.liger@assays[["RNA"]]@counts@Dimnames[[2]])
# colnames(cell_metadata) <- "barcode"
# 
# New_matrix <- merge.seurat.liger@assays[["RNA"]]@counts
# New_matrix <- New_matrix[rownames(gene_annotation), ]
# expression_matrix <- New_matrix
```


```{r}
##########
cds_from_seurat <- new_cell_data_set(expression_matrix,
                                     cell_metadata = cell_metadata,
                                     gene_metadata = gene_annotation)

### Construct and assign the made up partition

recreate.partition <- c(rep(1, length(cds_from_seurat@colData@rownames)))
names(recreate.partition) <- cds_from_seurat@colData@rownames
recreate.partition <- as.factor(recreate.partition)

cds_from_seurat@clusters@listData[["UMAP"]][["partitions"]] <- recreate.partition


### Assign the cluster info

list_cluster <- merge.seurat.liger@meta.data$labeled.ident2
names(list_cluster) <- merge.seurat.liger@assays[["RNA"]]@data@Dimnames[[2]]

cds_from_seurat@clusters@listData[["UMAP"]][["clusters"]] <- list_cluster


### Could be a space-holder, but essentially fills out louvain parameters

cds_from_seurat@clusters@listData[["UMAP"]][["louvain_res"]] <- "NA"


### Assign UMAP coordinate
cds_from_seurat@int_colData@listData$reducedDims[["UMAP"]] <-merge.seurat.liger@reductions[["umap"]]@cell.embeddings


### Assign feature loading for downstream module analysis

cds_from_seurat@preprocess_aux$gene_loadings <- merge.seurat.liger@reductions[["iNMF_raw"]]@feature.loadings


### Learn graph, this step usually takes a significant period of time for larger samples


#cds_from_seurat <- learn_graph(cds_from_seurat, use_partition = T, close_loop = FALSE,
#                               learn_graph_control = list(ncenter=100)) 

cds_from_seurat <- learn_graph(cds_from_seurat, use_partition = T, close_loop = FALSE)

#ncenter for 'branchiness'


### Plot cluster info with trajectory

clus <- plot_cells(cds_from_seurat, 
                color_cells_by = 'cluster',
                label_groups_by_cluster=FALSE,
                label_leaves=FALSE,
                label_branch_points=FALSE)
clus

#saveRDS(cds_from_seurat, file = "./seurat.objects/LIGERIntegration.cds.rds")

grep("CCCAATCAGATGCCAG-1", colnames(merge.seurat.liger))

root_cell_list <- c("GCTCCTACATACAGCT-1")
root_cell_list <- c("GCTCCTACATACAGCT-1", "CCCAATCAGATGCCAG-1")


cds_from_seurat <- order_cells(cds_from_seurat, root_cells = root_cell_list) #GUI does not work with modified pipeline right now



ptime <- plot_cells(cds_from_seurat, 
                color_cells_by = 'pseudotime',
                cell_size = 1,
                trajectory_graph_segment_size = 1.5,
                label_groups_by_cluster=FALSE,
                label_cell_groups=FALSE,
                label_leaves=FALSE,
                label_branch_points=FALSE)
ptime


cds_from_seurat_pr_test_res = graph_test(cds_from_seurat, neighbor_graph="principal_graph", cores=4)
pr_deg_ids = row.names(subset(cds_from_seurat_pr_test_res, q_value < 0.05))
  
gene_list_1 <- pr_deg_ids[1:4]

hgenes <- plot_cells(cds_from_seurat, 
                  genes = gene_list_1,
                  show_trajectory_graph=FALSE,
                  label_cell_groups=FALSE,
                  label_branch_points = FALSE,
                  label_leaves=FALSE)
hgenes 


PGAM_genes <- c("CCL2", "CCL3", "CCL4", "IL1B")
PGAM_lineage_cds <- cds_from_seurat[rownames(merge.seurat.liger) %in% PGAM_genes, PGAM.cells]

plot_genes_in_pseudotime(PGAM_lineage_cds,
                         min_expr=0.5)

```

#IDH Mutant cells seem to stretch everything. Subset them out, rerun iNMF for Monocle3
```{r}
Idents(merge.seurat.liger)<-"orig.sample"
cells.in<-WhichCells(merge.seurat.liger, idents = c('Huang', 'Yuan',
                                                    'Darmanis', 'Laffy', 'Sankowski'))

merge.seurat.liger.sub<-subset(merge.seurat.liger, idents = c('Huang', 'Yuan',
                                                    'Darmanis', 'Laffy', 'Sankowski'))

merge.seurat.liger.sub <- FindVariableFeatures(merge.seurat.liger.sub, selection.method = "vst", nfeatures = 2000)
merge.seurat.liger.sub <- ScaleData(merge.seurat.liger.sub, split.by = "orig.ident", do.center = FALSE)

#k=12 feat = 2200 neigh = 10 dist = 0.6 - looks eh
#k=10 feat = 2000 neigh = 30 dist = 0.3 - really bad
#k=20 feat = 2000 neigh = 30 dist = 0.3 - 

merge.seurat.liger.sub <- RunOptimizeALS(merge.seurat.liger.sub, k = 20, lambda = 5, split.by = "orig.ident")

merge.seurat.liger.sub <- RunQuantileAlignSNF(merge.seurat.liger.sub, split.by = "orig.ident")

merge.seurat.liger.sub <- RunUMAP(merge.seurat.liger.sub, 
                                      dims = 1:ncol(merge.seurat.liger.sub[["iNMF"]]), reduction = "iNMF",
                                      n.neighbors = 10, min.dist = 0.3)

DimPlot(merge.seurat.liger.sub, split.by = "labeled.ident2", ncol=4)
DimPlot(merge.seurat.liger.sub, group.by = "labeled.ident2", ncol=4)
```

#Diffusion Map

```{r}

library(destiny)
library(DESeq2)

liger.sub<-subset(merge.seurat.liger, cells= WhichCells(merge.seurat.liger, idents = c("Homeostatic Microglia",
                                                                                "Repopulating Microglia",
                                                                                "PGAM", "Sankowski_Aged", 
                                                                                "Tumor core GAM")))
liger.sub <- NormalizeData(liger.sub)
expression_matrix<-as.matrix(liger.sub@assays$RNA@data)

#remove zero variance genes
expression_matrix<-expression_matrix[apply(expression_matrix, 1, var) != 0, ]

#transform
expression_matrix<-log10(expression_matrix+1)

plotDensity(expression_matrix)

expression_matrix<-(t(expression_matrix))
expression_matrix<-as.ExpressionSet(as.data.frame(expression_matrix))

expression_matrix@phenoData$Cell<-liger.sub@meta.data$labeled.ident2

sigmas <- find_sigmas(expression_matrix, verbose = FALSE)
optimal_sigma(sigmas)

dif<-DiffusionMap(expression_matrix, n_pcs = 50, sigma = 8.237048)


sigmas <- find_sigmas(liger.sub@reductions$iNMF_raw@cell.embeddings, verbose = FALSE)
optimal_sigma(sigmas)

dif<-DiffusionMap(liger.sub@reductions$iNMF_raw@cell.embeddings)

#saveRDS(dif, file ="./seurat.objects/diffusion_dif.rds")

#all(rownames(dif@eigenvectors)== rownames(liger.sub@reductions$iNMF_raw@cell.embeddings))

Cells<-liger.sub@meta.data$labeled.ident2
ggplot(data = dif, aes(x=dif$DC1, y = dif$DC2))+
  geom_point(aes(color = Cells))

plot(dif)

plot(eigenvalues(dif), ylim = 0:1, pch = 20,
xlab = 'Diffusion component (DC)', ylab = 'Eigenvalue')


# Prepare a dataframe for cell plotting
plotting.data <- data.frame('DC1'= dif$DC1, 'DC2' = dif$DC2, 'DC3' = dif$DC3, 
                               'Cell' = Cells)

plot_ly(plotting.data, x = ~DC1, y = ~DC2, z = ~DC3, color = ~Cells, 
        size = 1) 

```

```{r}
library(destiny)

liger.sub<-subset(merge.seurat.liger, cells= WhichCells(merge.seurat.liger, 
                                                        idents = c("Homeostatic Microglia",
                                                                   "Repopulating Microglia",
                                                                   "PGAM", "Sankowski_Aged",
                                                                   "Tumor core GAM")))

dif<-DiffusionMap(liger.sub@reductions$iNMF_raw@cell.embeddings, n_pcs = 10,
                  k = 3300, verbose = TRUE)

#all(rownames(dif@eigenvectors) == colnames(liger.sub))

Cells<-liger.sub@meta.data$labeled.ident2

ggplot(data = dif, aes(x=dif$DC1, y = dif$DC2))+
  geom_point(aes(color = Cells))

#3D plot_ly
plotting.data <- data.frame('DC1'= dif$DC1, 'DC2' = dif$DC2, 'DC3' = dif$DC3, 
                               'Cell' = Cells)

plot_ly(plotting.data, x = ~DC1, y = ~DC2, z = ~DC3, color = ~Cells, 
        size = 1, colors = cols) 

```

#RACEID/StemID

```{r}




```
